### DISCUSSION TITLE
skipper supports forward-auth via webhook filter

### DISCUSSION THREAD

**Discussion by szuecs** [authelia/authelia#5200] [Ideas]
https://opensource.zalando.com/skipper/reference/filters/#webhook is a simple integration, that should be able to do the auth via [authelia](https://github.com/authelia/authelia) project as far as I understand.
https://www.authelia.com/docs/home/supported-proxies.html does not show it as supported and I guess we can check all the columns.

**Comment by james-d-elliott** [comment]
What headers does the webhook filter send (i.e. without configuration)?

**Comment by szuecs** [comment]
It copies all headers from original request to the target url specified in the filter, for example config: `webhook("https://authelia.foo.com")`.

**Comment by james-d-elliott** [comment]
Does skipper automatically add X-Forwarded-For, X-Forwarded-Proto, X-Fowarded-Host and X-Forwarded-Uri headers? or the X-Original-URL header? Or is this something that can be added? It would also be nice to add the X-Fowarded-Method header for full comparability before the request makes it to the webhook.

**Comment by james-d-elliott** [comment]
https://github.com/authelia/authelia/pull/3475/commits/7557bdb5e64193f304ce9db72cdf7461a6a02b52

**Comment by szuecs** [comment]
There are two option to influence the `X-Forwarded-*` headers:

```
  -forwarded-headers value
        comma separated list of headers to add to the incoming request before routing
        X-Forwarded-For sets or appends with comma the remote IP of the request to the X-Forwarded-For header value
        X-Forwarded-Host sets X-Forwarded-Host value to the request host
        X-Forwarded-Port=<port> sets X-Forwarded-Port value
        X-Forwarded-Proto=<http|https> sets X-Forwarded-Proto value
  -forwarded-headers-exclude-cidrs value
        disables addition of forwarded headers for the remote host IPs from the comma separated list of CIDRs
```

As client example call I do:
```
% curl -v -H"X-Foo: Bar Baz" http://127.0.0.1:9090/foo
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 9090 (#0)
> GET /foo HTTP/1.1
> Host: 127.0.0.1:9090
> User-Agent: curl/7.49.0
> Accept: */*
> X-Foo: Bar Baz
>
< HTTP/1.1 401 Unauthorized
< Server: Skipper
< Www-Authenticate: 127.0.0.1
< Date: Wed, 08 Jun 2022 08:10:46 GMT
< Transfer-Encoding: chunked
<
* Connection #0 to host 127.0.0.1 left intact
```

Example without settings, no `X-Forwarded-*` headers are set:
```
% ./bin/skipper -inline-routes '* -> webhook("http://127.0.0.1:12345/api") -> <shunt>'

# webhook
% nc -l 12345
GET /api HTTP/1.1
Host: 127.0.0.1:12345
User-Agent: curl/7.49.0
Accept: */*
X-Foo: Bar Baz
Accept-Encoding: gzip
```

Running skipper with these flags:
```
% ./bin/skipper -inline-routes '* -> webhook("http://127.0.0.1:12345/api") -> <shunt>' -forwarded-headers="X-Forwarded-For,X-Forwarded-Host"

# webhook
% nc -l 12345
GET /api HTTP/1.1
Host: 127.0.0.1:12345
User-Agent: curl/7.49.0
Accept: */*
X-Foo: Bar Baz
X-Forwarded-For: 127.0.0.1
X-Forwarded-Host: 127.0.0.1:9090
Accept-Encoding: gzip
```

Example with all set:
```
% ./bin/skipper -inline-routes '* -> webhook("http://127.0.0.1:12345/api") -> <shunt>' -forwarded-headers="X-Forwarded-For,X-Forwarded-Host,X-Forwarded-Port=443,X-Forwarded-Proto=https"

# webhook
% nc -l 12345
GET /api HTTP/1.1
Host: 127.0.0.1:12345
User-Agent: curl/7.49.0
Accept: */*
X-Foo: Bar Baz
X-Forwarded-For: 127.0.0.1
X-Forwarded-Host: 127.0.0.1:9090
X-Forwarded-Port: 443
X-Forwarded-Proto: https
Accept-Encoding: gzip
```

What should be passed in X-Forwarded-Uri  or X-Original-URL?
`/foo` or `127.0.0.1:9090/foo` ?

One important aspect is that skipper mimics the behavior of AWS Application Load Balancer, which appends the client connecting to the proxy to the X-Forwarded-For header.

**Comment by james-d-elliott** [comment]
> What should be passed in X-Forwarded-Uri or X-Original-URL?
> `/foo` or `127.0.0.1:9090/foo` ?

The URI / Path for `X-Forwarded-Uri`. So that when you make the request to `https://app.example.com/my/path` and the webhook makes a request to `http://authelia:9091/api/verify` we can determine the path is `/my/path`, without it we'd get `/api/verify` which is not overly useful. `X-Original-URL` should be the absolute URL entered into the browser. `X-Forwarded-Method` contains the HTTP request method verb.

**Comment by szuecs** [comment]
@james-d-elliott sounds like a great feature to add  X-Forwarded-Uri, X-Original-URL and X-Forwarded-Method on our side.

**Comment by szuecs** [comment]
After [PR](https://github.com/zalando/skipper/pull/2013) merge all of the X headers will be supported (release is automated).
```
GET /api HTTP/1.1
Host: 127.0.0.1:12345
User-Agent: curl/7.49.0
Accept: */*
X-Foo: Bar Baz
X-Forwarded-For: 127.0.0.1
X-Forwarded-Host: 127.0.0.1:9090
X-Forwarded-Method: GET
X-Forwarded-Port: 443
X-Forwarded-Proto: https
X-Forwarded-Uri: /foo
X-Original-Url: https://127.0.0.1:9090/foo
Accept-Encoding: gzip
```

Is It required to have `X-Original-URL` or is `X-Original-Url` good enough?
The first would make some hoops required to make it work

**Comment by nightah** [comment]
You only need one of either X-Original-URL or the set of X-Forwarded-* headers.

Except if you use X-Original-URL you also need X-Forwarded-Method if performing method filtering.

**Comment by james-d-elliott** [comment]
Oh I'm sorry @szuecs didn't realize you were one of the devs. Yeah as @nightah mentioned, we don't need both. Probably best to ignore the `X-Original-URL` in that case. Traefik and Caddy only use `X-Forwarded-*` by default. If you need some specifics in how both of those operate I'm more than willing to shed some light on it, however I feel like the existing implementation with the additional headers roughly matches.

The only thing that is different from memory is that any response returned not in the 2xx range by traefik and caddy is forwarded / copied to the browser. Requests that are XHR's (for apps not using the new fetch API) can be a bit of a pain to redirect if this isn't the case, as generally they redirect more successfully when the response code is a 401 with a Location header. Though in my recent investigation with Caddy I can't say I can find any documentation on this anymore.

Either way maybe there is an option in the future to make another filter that borrows most of its behavior from the webhook filter and forwards the request and cancels the request to the backend if the response is not a 2xx?

**Comment by szuecs** [comment]
The webhook filter in skipper won't forward a request to backends in case webhook responses with >=300 status code, https://github.com/zalando/skipper/blob/master/filters/auth/webhook.go#L127

I implemented that users can choose to get `X-Original-URL` or `X-Forwarded-*`, so flexible enough for them to use authelia.

**Comment by james-d-elliott** [comment]
> The webhook filter in skipper won't forward a request to backends in case webhook responses with >=300 status code, https://github.com/zalando/skipper/blob/master/filters/auth/webhook.go#L127
> 
> I implemented that users can choose to get `X-Original-URL` or `X-Forwarded-*`, so flexible enough for them to use authelia.

I got that part. What I meant specifically is both Caddy and Traefik forward the auth servers response (authelia) to the browser in any instance which isn't a 2xx code or 300 code provided there are no errors such as gateway timeouts etc. It's not a breaker, it would just be a nice to have since it'd give Authelia control over how the process occurs (i.e. we redirect them to the Authelia portal).

**Comment by szuecs** [comment]
True, I guess it makes sense for UI flows. Would this be fine to change the condition that:
`<300` forwarded to backend
`>=300` && `<400` send response directly to the client
`>=400` stop processing as it is

Like this you can redirect as you like and it would not be such a big change for us.

**Comment by james-d-elliott** [comment]
That would technically work, and would be a good step in the short term. 

It is my experience however that the legacy XHR requests will swallow redirects, and the only workaround is sending status 401 with a Location header. I'll try to dig up some information to confirm this point but struggled to find it recently when working with Caddy.

What I was thinking was two versions of the same filter which are marginally different. Most of the underlying machinery can be 100% the same, and then potentially via a decorator or strategy pattern (or potentially just via bool or int fields which define the actual behavior) you could alter the behavior in one direction or the other.

**Comment by szuecs** [comment]
I guess we can also copy the Location header on 401.

**Comment by james-d-elliott** [comment]
Think that'll easily suffice. I personally believe it would be less intrusive to users to make it similar to another implementation but it will work for our use case. 

I updated our docs, also will aim to setup an integration suite for skipper in the coming weeks so we can adequately test it and come up with some official docs. Also sorry for my delays, my time is very limited lately.

If you had a basic docker-compose example of usage with a simple docker app (or advanced depending on what is available) that would be nice. I did have a brief look but nothing stood out to me. If you don't that's fine to and I can just figure it out.

**Comment by james-d-elliott** [comment]
If you'd like to collaborate on this let me know, I couldn't really figure out from the docs but I don't have a lot of free time.

**Comment by james-d-elliott** [comment]
I'm going to remove Skipper for now since it doesn't seem like support is possible.

**Reply by szuecs** [reply]
Yeah fair enough, sorry I didn't find the time to follow up. Maybe I will come back at a later time when I have more time to explore this.
